engine/access/ingestion/collection_syncer.go

```diff
@@ -0,0 +1,398 @@
+//
+// Expected errors during normal operations:
+// - storage.ErrNotFound if no block is found in storage for the given height
+func (s *CollectionSyncer) findMissingCollectionsAtHeight(height uint64) ([]*flow.CollectionGuarantee, error) {
+	block, err := s.blocks.ByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve block by height %d: %w", height, err)
+	}
+
+	var missingCollections []*flow.CollectionGuarantee
+	for _, guarantee := range block.Payload.Guarantees {
+		inStorage, err := s.isCollectionInStorage(guarantee.CollectionID)
+		if err != nil {
+			return nil, err
+		}
+
+		if !inStorage {
+			missingCollections = append(missingCollections, guarantee)
+		}
+	}
+
+	return missingCollections, nil
+}
+
+// isCollectionInStorage checks whether the given collection is present in local storage.
+// No errors expected during normal operation
+func (s *CollectionSyncer) isCollectionInStorage(collectionID flow.Identifier) (bool, error) {
+	_, err := s.collections.LightByID(collectionID)
+	if err == nil {
+		return true, nil
+	}
+
+	if errors.Is(err, storage.ErrNotFound) {
+		return false, nil
+	}
+
+	return false, fmt.Errorf("failed to retrieve collection %s: %w", collectionID.String(), err)
+}
+
+// RequestCollectionsForBlock conditionally requests missing collections for a specific block height,
+// skipping requests if the block is already below the known full block height.
+func (s *CollectionSyncer) RequestCollectionsForBlock(height uint64, missingCollections []*flow.CollectionGuarantee) {
+	// skip requesting collections, if this block is below the last full block height.
+	// this means that either we have already received these collections, or the block
+	// may contain unverifiable guarantees (in case this node has just joined the network)
+	if height <= s.lastFullBlockHeight.Value() {
+		s.logger.Info().
+			Msg("skipping requesting collections for finalized block as its collections have been already retrieved")
+		return
+	}
+
+	s.requestCollections(missingCollections, false)
+}
+
+// requestCollections registers collection download requests in the requester engine,
+// optionally forcing immediate dispatch.
+func (s *CollectionSyncer) requestCollections(missingCollections []*flow.CollectionGuarantee, immediately bool) {
+	for _, collection := range missingCollections {
+		guarantors, err := protocol.FindGuarantors(s.state, collection)
+		if err != nil {
+			// failed to find guarantors for guarantees contained in a finalized block is fatal error
+			s.logger.Fatal().Err(err).Msgf("could not find guarantors for guarantee %v", collection.ID())
+		}
+		s.requester.EntityByID(collection.ID(), filter.HasNodeID[flow.Identity](guarantors...))
+	}
+
+	if immediately {
+		s.requester.Force()
+	}
+}
+
+// updateLastFullBlockHeight updates the next highest block height where all previous collections have been indexed.
+// No errors expected during normal operation.
```

Comment:
let's be consistent with the formatting
```suggestion
//
// No errors expected during normal operation.
```